### PR TITLE
expression: handle corrupted length in uncompress builtin function

### LINKED ISSUE #8578: builtin uncompress function ignored the four bytes length
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?

```
mysql> select uncompress(unhex('02000000789CCB48CDC9C95728CF2FCA4901001A0B045D'));
```

2. What did you expect to see?
 
```
mysql> select uncompress(unhex('02000000789CCB48CDC9C95728CF2FCA4901001A0B045D'));
+---------------------------------------------------------------------+
| uncompress(unhex('02000000789CCB48CDC9C95728CF2FCA4901001A0B045D')) |
+---------------------------------------------------------------------+
| NULL                                                                |
+---------------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+--------------------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                                          |
+---------+------+--------------------------------------------------------------------------------------------------+
| Warning | 1258 | ZLIB: Not enough room in the output buffer (probably, length of uncompressed data was corrupted) |
+---------+------+--------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

3. What did you see instead?
```
+---------------------------------------------------------------------+
| uncompress(unhex('02000000789CCB48CDC9C95728CF2FCA4901001A0B045D')) |
+---------------------------------------------------------------------+
| hello world                                                         |
+---------------------------------------------------------------------+
1 row in set (0.01 sec)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
tidb_version(): Release Version: v2.1.0-rc.2-10-gb30dbd0
Git Commit Hash: b30dbd06acaab088a1e8fa004e189c651ca08bf1
Git Branch: master
UTC Build Time: 2018-09-17 01:20:48
GoVersion: go version go1.11 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
```

5. more info
This is found when I try to implement builtin compress and uncompress function in  TiKV (https://github.com/tikv/tikv/pull/3856)
the first four bytes stored the  length of the uncompress string, if it's less than the corrected length (maybe corrupted), MySQL  return Null， TiDB just ignore it.

### PR DESCRIPTION
**Pull request by ziyi-yan**
### What problem does this PR solve? 
fixes #8578 

### What is changed and how it works?
Returns nil with [warning](https://dev.mysql.com/doc/refman/5.6/en/server-error-reference.html#error_er_zlib_z_buf_error) when the length in first four bytes is less than the length of the returned string.

While fixing this issue, I [change](https://github.com/pingcap/tidb/pull/8586/files#diff-657d3eaf9cd8c201d0d716f73eb1892dL34) the error class of `errZlibZData` from `ClassTypes` to `ClassExpression` in order to display their SQL error codes correctly.

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8586)

### CODE REVIEW

File: expression/errors.go [hunk 1]
```diff
@@ -31,7 +31,8 @@ var (
 
 	// All the un-exported errors are defined here:
 	errFunctionNotExists             = terror.ClassExpression.New(mysql.ErrSpDoesNotExist, mysql.MySQLErrName[mysql.ErrSpDoesNotExist])
-	errZlibZData                     = terror.ClassTypes.New(mysql.ErrZlibZData, mysql.MySQLErrName[mysql.ErrZlibZData])
+	errZlibZData                     = terror.ClassExpression.New(mysql.ErrZlibZData, mysql.MySQLErrName[mysql.ErrZlibZData])
```
Comment (ziyi-yan) [inline expression/errors.go]: While fixing this issue, I change the error class of `errZlibZData` from `ClassTypes` to `ClassExpression` in order to display their SQL error codes correctly.

**Comment by ziyi-yan** [comment]
PTAL @zimulala

**Comment by ziyi-yan** [comment]
ping @zimulala @shenli

**Comment by ziyi-yan** [comment]
Are there any problems blocking this PR?

**Comment by zz-jason** [comment]
@ziyi-yan Could you help us cherry-pick this commit to release-2.1 and release-2.0?

**Comment by ziyi-yan** [comment]
@zz-jason I would love to do it. But what am I supposed to do exactly? (I am not an experienced contributor yet.)
I might run `git cherry-pick <this-commit-id>` to branch `release-2.0` and `release-2.1` and make two PRs to these branches. Is that correct?

**Comment by zz-jason** [comment]
> @zz-jason I would love to do it. But what am I supposed to do exactly? (I am not an experienced contributor yet.)
> I might run `git cherry-pick <this-commit-id>` to branch `release-2.0` and `release-2.1` and make two PRs to these branches. Is that correct?

Yes, that's right. Feel free to ask if you encountered any problem. 😄

**Pull request merged**